### PR TITLE
[M68k] Introduce more MOVI cases

### PR DESCRIPTION
**Pull request by n8pjl**
Add three more special cases for loading registers with immediates.

The first allows values in the range of [-255, 255] to be loaded with MOVEQ, even if the register is more than 8 bits and the sign extention is unwanted. This is done by loading the bitwise complement of the desired value, then performing a NOT instruction on the loaded register.

This special case is only used when a simple MOVEQ cannot be used, and is only used for 32 bit data registers. Address registers cannot support MOVEQ, and the two-instruction sequence is no faster or smaller than a plain MOVE instruction when loading 16 bit immediates on the 68000, and likely slower for more sophisticated microarchitectures. However, the instruction sequence is both smaller and faster than the corresponding MOVE instruction for 32 bit register widths.

The second special case is for zeroing address registers. This simply expands to subtracting a register with itself, consuming one instruction word rather than 2-3, with a small improvement in speed as well.

The last special case is for assigning sign-extended 16-bit values to a full address register. This takes advantage of the fact that the movea.w instruction sign extends the output, permitting the immediate to be smaller. This is similar to using lea with a 16-bit address, which is not added in this patch as 16-bit absolute addressing is not yet implemented.

This is a v2 submission of #90817. It also creates a 'Data' test directory to better align with the backend's tablegen layout.

### CODE REVIEW

**Comment by n8pjl** [comment]
That was a small cleanup that I meant to do before submitting. I have no further changes to make.

**Comment by n8pjl** [comment]
ping @glaubitz @mshockwave

**Comment by glaubitz** [comment]
Thanks. I'll have a look although the final approval needs to come from @mshockwave or @0x59616e.

**Comment by 0x59616e** [comment]
I may not be the right person to review this because I've been away from llvm and m68k backend for a long time.

File: llvm/lib/Target/M68k/M68kInstrInfo.cpp [hunk 1]
```diff
@@ -360,18 +360,66 @@ bool M68kInstrInfo::ExpandMOVI(MachineInstrBuilder &MIB, MVT MVTSize) const {
...
-    Register SReg =
-        MVTSize == MVT::i32 ? Reg : Register(RI.getMatchingMegaReg(Reg, DR32));
-    assert(SReg && "No viable MEGA register available");
-
     MIB->setDesc(get(M68k::MOVQ));
     MIB->getOperand(0).setReg(SReg);
+
+    // Counter the effects of sign-extension with a bitwise not.
+    // This is only faster and smaller for 32 bit values.
+  } else if (DR32->contains(Reg) && std::abs(Imm) <= 255) {
```
Comment (mshockwave) [inline llvm/lib/Target/M68k/M68kInstrInfo.cpp:381]: could we use `llvm::isUInt<8>(Imm)` here?

File: llvm/lib/Target/M68k/M68kInstrInfo.cpp [hunk 2]
```diff
@@ -360,18 +360,66 @@ bool M68kInstrInfo::ExpandMOVI(MachineInstrBuilder &MIB, MVT MVTSize) const {
...
+        .addReg(SReg, RegState::Undef);
+
+    MIB->removeFromParent();
+
+    // movea.w implicitly sign extends to the full register width,
+    // so exploit that if the immediate fits in the correct range.
+    //
+    // TODO: use lea imm.w, %an for further constants when 16-bit
+    // absolute addressing is implemented.
+  } else if (AR32->contains(Reg) && Imm >= -32768 && Imm <= 32767) {
```
Comment (mshockwave) [inline llvm/lib/Target/M68k/M68kInstrInfo.cpp:413]: ditto, `llvm::isUInt<16>(Imm)`?

**Review by mshockwave — COMMENTED** [review]
My sincere apology for the turnaround time. Generally looks good to me w/ minor comments.

File: llvm/lib/Target/M68k/M68kInstrInfo.cpp (hunk 1 above)
Comment (n8pjl) [reply llvm/lib/Target/M68k/M68kInstrInfo.cpp]: I didn't know about that. It's cleaner and works the same.

**Review by mshockwave — APPROVED** [review]
LGTM thanks

**Comment by n8pjl** [comment]
Do you mind committing this PR? I don't have access.

**Pull request merged**